### ISSUE TITLE
Scenes should track cumulative time since starting

### ISSUE THREAD

**Issue by eulertour** [ManimCommunity/manim#851]
## Description of proposed feature
Scenes should store the total time since starting and make this value accessible via a public API.

## How can the new feature be used?
There isn't currently a way to update mobjects based on the cumulative elapsed time since a given start point. Time based updaters only pass `dt`, the incremented time, which leaves the user to add that value to a running total manually in order to track cumulative time. This makes things like animating a continually growing angle unnecessarily complex.

It may be best to also add a new Clock class which inherits from ValueTracker and has APIs like `Scene.get_clock()`, `Clock.start()`, and `Clock.stop()`.

**Comment by leotrs** [comment]
Agreed with the general point.

One nitpick: I would rather see a complete overhaul of ValueTracker before we inherit anything from it. Every single use of ValueTracker I've seen seems unnecessarily clunky...

**Comment by eulertour** [comment]
What sort of clunkiness are you trying to avoid?

**Comment by NeoPlato** [comment]
I support this venture. Seems helpful for the `add_sound` method in the scenes.

So there's the basic part: using it the first time is an error off the bat:

```python
class AudioTest(Scene):
    def construct(self):
        group_dots = VGroup(*[Dot() for _ in range(3)])
        group_dots.arrange_submobjects(RIGHT)
        for dot in group_dots:
            self.add_sound("generic_sound.wav")
            self.play(FadeIn(dot))
            self.wait()


  File "C:\Users\User\ManimCommunity\manim\scene\scene.py", line 861, in add_sound
    time = self.time + time_offset
AttributeError: 'AudioTest' object has no attribute 'time'
```

So I take care of that by changing `self.time` to `self.renderer.time` here:

![Yes](https://cdn.discordapp.com/attachments/581738732646957057/787693652699774976/unknown.png)

The animation actually plays well, with the sound in sync with the dots ()

[The video, I think. Will unfortunately download if you click](https://cdn.discordapp.com/attachments/581738732646957057/787695543463444480/AudioTest.mp4)

Good tidings and fortune right? A man does a rerun:

![From trusty Visual Studio Code](https://cdn.discordapp.com/attachments/581738732646957057/787700143260434432/unknown.png)

[More links you can trust*](https://cdn.discordapp.com/attachments/581738732646957057/787707738599063642/AudioTest)

So the sound is not in sync. Hence caching affects the time count. How exactly? I spoke with my debugger:

### 1. Without caching. Note the number of frames on the left hand side

> ![image](https://user-images.githubusercontent.com/67624028/102625946-3c9fd500-4157-11eb-9b7d-6e18fdee4d56.png)

> ![image](https://user-images.githubusercontent.com/67624028/102626139-7a9cf900-4157-11eb-8cd9-b6c72ae0bd30.png)

> ![image](https://user-images.githubusercontent.com/67624028/102626268-a15b2f80-4157-11eb-8888-234c9ea57737.png)

> 60 frames. All accounted for, right?


### 2. With caching

> ![image](https://user-images.githubusercontent.com/67624028/102626718-39f1af80-4158-11eb-8c93-e9c2517eae1c.png)

> ![image](https://user-images.githubusercontent.com/67624028/102626734-41b15400-4158-11eb-8733-d4dc723885dd.png)

> ![image](https://user-images.githubusercontent.com/67624028/102626768-4bd35280-4158-11eb-8404-1999b35a9365.png)

Considering `add_sound` directly uses that attribute to figure out where to place the sound, a separate class for this independent of caching effects is most welcome

**Comment by eulertour** [comment]
@NeoPlato is right, this is already being done in `Scene.renderer.time`, so the solution shouldn't be any more complicated than adding a property to Scene to read the value from the renderer.

**Comment by eulertour** [comment]
Scene caching does change it, but if the scene is cached anyway maybe that isn't a problem?

**Comment by NeoPlato** [comment]
> @NeoPlato is right, this is already being done in `Scene.renderer.time`, so the solution shouldn't be any more complicated than adding a property to Scene to read the value from the renderer.

Would that ensure the frames get updated correctly even with caching?

**Comment by eulertour** [comment]
The renderer time is updated in `CairoRenderer.add_frame()` and that's called from `Scene.play_internal()` which is skipped during caching, so unless this is handled somewhere else in the caching logic frames that depend on the cumulative time won't work properly if they follow a cached animation.

https://github.com/ManimCommunity/manim/blob/master/manim/renderer/cairo_renderer.py#L144
https://github.com/ManimCommunity/manim/blob/master/manim/scene/scene.py#L826

**Comment by NeoPlato** [comment]
> that's called from `Scene.play_internal()` which is skipped during caching

Oh it skips the entire thing? That makes sense.

So you would have a separate construct that updates the frame from the `Scene` class independent of the caching

**Comment by Hari-07** [comment]
I'd like to take up this issue but since I havent worked on this project before a few pointers might help to start with this issue

**Comment by leotrs** [comment]
@Hari-07 thank you for volunteering! What kind of pointers are you looking for? If you start your work and come across a concrete doubt, don't hesitate to ask here or on our discord.

**Comment by huguesdevimeux** [comment]
Hi @Hari-07 , are you still willing to work on this? If you don't manifest, I think I'll do it. 

As @eulertour  Taking the elapsed time from the renderer is a bad idea when a scene is skipped for some reason.

Instead, we could when a scene is skipped add its time to an internal clock of the Scene.

**Comment by Hari-07** [comment]
Yeah sure you can take it, I got engrossed in other side projects and
forgot about this to be honest
